### DISCUSSION TITLE
Implicit "this"?

### DISCUSSION THREAD

**Discussion by lerno** [c3lang/c3c#439] [Polls]
Possibly I would introduce an implicit `this` parameter for method functions, and I'd like to hear what you think about it. Below you see three possible versions: A (current), B (implicit), C (explicit but without the type). I'm still collecting opinions here, so nothing is decided yet.

```c
fn void test(Foo* f) {
  printf("%d", f.bar(123));
}
// A explicit "this" parameter:
fn int Foo.bar(Foo* foo, int x) {
  return foo.z * x;
}
// B implicit "this" parameter (but still explicit use)
fn int Foo.bar(int x) {
  return this.z * x;
}
// C explicit "this" parameter but without type:
fn int Foo.bar(foo, int x) {
  return foo.z * x;
}
```

**Comment by zweimach** [comment]
(D) Explicit but with Go-style method syntax

```c
// D explicit but with Go-style method syntax:
fn int (Foo* foo) bar(int x) {
    return foo.z * x;
}
```

**Reply by lerno** [reply]
> (D) Explicit but with Go-style method syntax

I'm afraid that I find the Go style syntax extremely hard to visually parse and also feels very different from the rest of the syntax.

**Comment by data-man** [comment]
~(D)~ (E)
this -> [$self](https://wiki.freepascal.org/Self) like FreePascal:
[Programming Using Objects](https://wiki.freepascal.org/Programming_Using_Objects)
[Programming Using Objects Page 2](https://wiki.freepascal.org/Programming_Using_Objects_Page_2)

```rust
fn void Point.inc(int sx, int sy) {
  with $self {
      x += sx;
      y += sy;
  }
}
```

**Reply by lerno** [reply]
The suggestion is not for using member access without `this`. The `$` prefix is reserved for compile time variables.

**Comment by 00shiv** [comment]
What about option A, but allow dot calls on any value of the first argument type?
```c
fn void test(Foo* f) {
  printf("%d", f.bar(123));
}
// A explicit "this" parameter; but can call it without first argument via dot notation
fn int bar(Foo* foo, int x) {
  return foo.z * x;
}
```

**Reply by lerno** [reply]
That would be function overloading, which C3 doesn't have.

**Comment by lerno** [comment]
This is long since resolved. So closing it.
